### PR TITLE
Remove the Array pinning addressor and all the pinning instructions in SIL

### PR DESCRIPTION
**Pull request by eeckstein**
Pinning was used to handle simultaneous accesses to arrays, like for swap(&a[x], &a[y]).
With the rule of exclusivity this is not allowed anymore. And as we will have exclusivity checking in the runtime, we are enforcing this rule. So we don't need the pinning mechanism anymore.

This PR removes the pinning addressors, the pinning built-ins and the corresponding SIL instructions.

To do: remove the runtime entry points.

rdar://problem/35401528

This change causes a few benchmark regressions because of SILGen inefficiencies/bugs: https://bugs.swift.org/browse/SR-8621

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

File: docs/ABI/Mangling.rst [hunk 1]
```diff
@@ -232,7 +232,7 @@ Entities
   ADDRESSOR-KIND ::= 'u'                     // unsafe addressor (no owner)
   ADDRESSOR-KIND ::= 'O'                     // owning addressor (non-native owner)
   ADDRESSOR-KIND ::= 'o'                     // owning addressor (native owner)
-  ADDRESSOR-KIND ::= 'p'                     // pinning addressor (native owner)
+  ADDRESSOR-KIND ::= 'p'                     // pinning addressor (native owner), not used anymore
```
Comment (slavapestov) [inline docs/ABI/Mangling.rst]: Since addressors never had public symbols do we still need to document and demangle them?

Comment (eeckstein) [reply docs/ABI/Mangling.rst]: I think we don't need to. On the other hand it does not harm to be able to demangle Swift < 5 symbols, even if they are not public.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: docs/ABI/Mangling.rst (hunk 1 above)
Comment (slavapestov) [reply docs/ABI/Mangling.rst]: At the very least you can remove mangling from the ASTMangler, even if you intend on keeping the demangler and remangler support.

Comment (eeckstein) [reply docs/ABI/Mangling.rst]: I did

**Pull request merged**